graphistry/embed_utils.py

```diff
@@ -0,0 +1,153 @@
+import torch
+import torch.nn as nn
+from torch.utils.data import Dataset, DataLoader
+from tqdm import trange
+
+class HeterographEmbedModuleMixin(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+        self.protocol = {
+                'TransE': self.TransE,
+                'DistMult': self.DistMult,
+                'RotatE': self.RotatE
+        }
+
+    def embed(self, proto='TransE', d=128, gamma=12):
+
+        # initialize hparams (from arguments)
+        self.EPS = 2.0
+        self.d = d
+        self.emrange = gamma + self.EPS / d
+        
+        # initializing the model
+        em_mod = HeteroEmbed(len(self._nodes), self.EPS, d, self.emrange, self.protocol[proto])
+
+        # entity2id <- from graphistry graph
+        entity2id = dict()
+        src = set(self._edges[self._source]).union(set(self._edges[self._destination]))
+        for idx, s in enumerate(src):
+            entity2id[s] = idx
+
+        # relation2id <- from graphistry graph
+        relation2id = dict()
+        rel = set(self._edges[self._relation])
+        for idx, r in enumerate(rel):
+            relation2id[r] = idx
+
+        # build triplets (TODO: Efficiency)
+        triplets = []
+        src = list(self._edges[self._source])
+        dst = list(self._edges[self._destination])
+        relation = list(self._edges[self._relation])
+
+        for s, r, d in zip(src, relation, dst):
+            triplets.append([s, r, d])
+        # initialize the dataloader
+        dataset = EmbedDataset(entity2id, relation2id, triplets)
+        train_generator = DataLoader(dataset, batch_size=32) # need autoscale batch for gpu? [AUTOML]
+
+        # training loop
+        optim = torch.optim.Adam(em_mod.parameters(), lr=0.01)
+
+        # TODO: define epoch [AUTOML]
+
+        epoch = trange(1000, leave=True)
+        for _ in epoch:
+            loss_acc, c = 0, 0
+
+            #for _ in trange(1000):
+            em_mod.train()
+            for h, r, t in train_generator:
+
+                optim.zero_grad()
+
+                pos_triples = torch.stack((h, r, t), dim=1)
+
+                # negative sample generation
+                h_o_t = torch.randint(high=2, size=h.size())
+                random_nodes = torch.randint(high=len(entity2id), size=h.size())
+                neg_h = torch.where(h_o_t == 0, random_nodes, h)
+                neg_t = torch.where(h_o_t == 1, random_nodes, t)
+                neg_triples = torch.stack((neg_h, r, neg_t), dim=1)
+
+                loss = em_mod(pos_triples, neg_triples)
+                loss.mean().backward()
+                loss_acc += loss.mean().item()
+                c += 1
+                optim.step()
+
+            epoch.set_description(f"loss: {loss_acc/c}")
+            epoch.refresh()
+
+        self.node_em = em_mod.node_em.weight.detach().numpy()
+        self.edge_em = em_mod.edge_em.weight.detach().numpy()
+        return self.node_em, self.edge_em
```

Comment:
return a new graphistry instance, rather than vectors, so we can continue with the fluent api.

Reply:
so just like `_node_features`, we have `_node_embedding` with some variable dim, and `.embed()` can memoize on features, model -> embed?

Reply:
now it returns a new graphistry instance with `G_node_features = [G_old_node_features | G_relational_embed_features]`

Reply:
@silkspace should we concat or add `G._{node,edge}_embedding` ?